middleware/auth.go

```diff
@@ -0,0 +1,73 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gol4ng/httpware/v2"
+	"github.com/gol4ng/httpware/v2/auth"
+)
+
+// Authentication middleware delegate the authentication process to a authFunc configured
+func Authentication(options ...AuthOption) httpware.Middleware {
+	config := NewAuthConfig(options...)
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+			newCtx, err := config.authFunc(req)
+			if err != nil && config.errorHandler(err, writer, req) {
+				return
+			}
+
+			next.ServeHTTP(writer, req.WithContext(newCtx))
+		})
+	}
+}
+
+type authFunc func(req *http.Request) (context.Context, error)
+type errorHandler func(err error, writer http.ResponseWriter, req *http.Request) bool
+
+// AuthOption defines a interceptor middleware configuration option
+type AuthOption func(*AuthConfig)
+
+type AuthConfig struct {
+	authFunc     authFunc
+	errorHandler errorHandler
+}
+
+func (o *AuthConfig) apply(options ...AuthOption) {
+	for _, option := range options {
+		option(o)
+	}
+}
+
+func NewAuthConfig(options ...AuthOption) *AuthConfig {
+	opts := &AuthConfig{
+		authFunc:     DefaultAuthFunc,
+		errorHandler: DefaultErrorHandler,
+	}
+	opts.apply(options...)
+	return opts
+}
+
+func DefaultAuthFunc(req *http.Request) (context.Context, error) {
+	return auth.CredentialToContext(req.Context(), auth.FromHeader(req)()), nil
+}
+
+func DefaultErrorHandler(err error, writer http.ResponseWriter, _ *http.Request) bool {
+	http.Error(writer, err.Error(), http.StatusUnauthorized)
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication

An HTTP response having a `401` statusCode should also return a `WWW-Authenticate` HTTP header containing the available authorization methods for the requested resource 👍

Reply:
We ignore the type of authentication method was try so we cannot return the appropriate `WWW-Authenticate` value
It's currently a developer side override